### ISSUE TITLE
IDE0008 when combining "type apparent" expressions with null-coalescing operator

### ISSUE THREAD

**Issue by jimpaton-ufx** [dotnet/roslyn#48011]
With the following settings:
``` ini
csharp_style_var_for_built_in_types = false:warning
csharp_style_var_when_type_is_apparent = true:suggestion
csharp_style_var_elsewhere = false:warning
```
a false-positive IDE0008 is generated when combining two "type is apparent" expressions with the null-coalescing operator `??`, for example:

``` csharp
var myList = enumerableOfInt as List<int> ?? enumerableOfInt.ToList();
```
No warning is generated when using the expressions either side of the `??` in isolation.

**Version Used**: 
Microsoft Visual Studio Enterprise 2019
Version 16.7.3
VisualStudio.16.Release/16.7.3+30503.244
Microsoft .NET Framework
Version 4.8.04084

**Steps to Reproduce**:

1. Create a new console app project
2. Add a new .editorconfig file and ensure the options for `charp_style_var_*` are set as follows:
``` ini
[*.cs]
# var preferences
csharp_style_var_for_built_in_types = false:warning
csharp_style_var_when_type_is_apparent = true:suggestion
csharp_style_var_elsewhere = false:warning
```
3. Add the following code to `Main`
``` csharp
IEnumerable<int> empty = Enumerable.Empty<int>();

var emptyAsList = empty as List<int>;
var emptyToList = empty.ToList();
var emptyAsListOrToList = empty as List<int> ?? empty.ToList();
```

**Expected Behavior**:
No IDE0008, because `csharp_style_var_when_type_is_apparent  = true:suggestion`

**Actual Behavior**:
IDE0008 on third example of `var` because `csharp_style_var_elsewhere = false:warning`

**Comment by CyrusNajmabadi** [comment]
I don't believe the type is apparent there. Is requires descending into both sides of the expression, then mentally comparing both to know of they're the same. This is not apparent. 

In general we do not done into expressions on the way you're thinking about.

**Comment by jimpaton-ufx** [comment]
I'm sorry, but I disagree. Given the following lines of code:

``` csharp
var first = String.Empty;
var second = String.Empty;
var result = first ?? second;
```

with  `csharp_style_var_for_built_in_types` to `false:warning`, all three lines generate an IDE0008 warning, as expected. If you change value of the setting to `true:warning` then all three IDE0008 warnings disappear. My expectation is that the rules should behave in a consistent manner. If `var x = expressionA;` and `var y = expressionB;` don't generate warnings under a particular rule, then why should `var z = expressionA ?? expressionB;` be different.

**Comment by CyrusNajmabadi** [comment]
> My expectation is that the rules should behave in a consistent manner

There is no reason for this to be.  `csharp_style_var_for_built_in_types` depends only on the type of hte variable.  if it is `built in` (i.e. `int`, `string`, etc.) you can use 'var'.  that's the only thing that controls it.

`csharp_style_var_when_type_is_apparent` is entirely unrelated.  It says "you can use `var` if the type is apparent/obvious from the expression".  When you have `expr1 ?? expr2` the type is not apparent/obvious, so you cannot use 'var' here.

> then why should `var z = expressionA ?? expressionB`; be different?

You'd have to explain why `var z = expressionA ?? expressionB;` would be the same.  it's a `??` expression, nothing about it make the type `apparent`.  The same is true for things like `x ? y : z`.  Is the type of that apparent just by looking at it?  No.  So you can't use 'var' there.

**Comment by jimpaton-ufx** [comment]
I understand. My (I think reasonable) assumption was that both rules would be looking at the type of the RHS of the assignment - from a human being's point of view, I can't tell whether `var x` is a built-in type or not!
Thanks for clarifying.

**Comment by CyrusNajmabadi** [comment]
Yup yup. I'm going to close this as by design. Glad this could be cleared up for you.
